### PR TITLE
IOOP-16 Add interaction with io backend

### PR DESCRIPTION
**Pull request by datalek**

### CODE REVIEW

File: .env.default [hunk 1]
```diff
@@ -1,8 +1,9 @@
 SERVER_HOSTNAME="localhost"
-SERVER_PORT="3000"
+SERVER_PORT="3001"
```
Comment (datalek) [inline .env.default:2]: Is this change required due to the demo?

File: .env.default [hunk 2]
```diff
@@ -1,8 +1,9 @@
...
 LOG_LEVEL="debug"
 APPLICATION_NAME="io-openid-provider"
 REDIS_URL="redis://0.0.0.0:6379"
 REDIS_KEY_PREFIX="oidc:"
 TEST_CLIENT_ID="example-client"
-TEST_CLIENT_REDIRECT_URI="https://relying-party.io/callback"
+TEST_CLIENT_REDIRECT_URI="https://a604-93-43-201-145.ngrok.io/callback"
```
Comment (datalek) [inline .env.default:8]: Is this change required due to the demo?

File: src/oidcprovider/__tests__/router.test.ts [hunk 3]
```diff
@@ -50,7 +50,11 @@ describe("interactionLogic", () => {
     const detail = records.loginPromptDetail;
     const mockUserInfoClient = phonies.makeMockUserInfoClient();
     mockUserInfoClient.findUserByFederationToken.mockReturnValue(
-      TE.right({ id: cookieValue })
+      TE.right({
+        fiscalCode: cookieValue,
```
Comment (datalek) [inline src/oidcprovider/__tests__/router.test.ts:54]: Can we move this information to `records.ts`?

File: src/userinfo/ioUserInfoClient.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
+import * as TE from "fp-ts/lib/TaskEither";
+import * as f from "fp-ts/function";
+import nodeFetch from "node-fetch";
+import * as authClient from "../generated/clients/io-auth/client";
+import { UserIdentity } from "../generated/clients/io-auth/UserIdentity";
+import { Client, createClient } from "../generated/clients/io-auth/client";
+import * as I from ".";
+
+const toError = (type: I.ErrorType): I.UserInfoClientError => ({
```
Comment (datalek) [inline src/userinfo/ioUserInfoClient.ts:9]: What do you think to move this function on `<pr-home>/src/userinfo/index.ts` file?

File: src/userinfo/ioUserInfoClient.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      }),
+      TE.map(makeUserInfo)
+    );
+
+const makeIOUserInfoClient = (client: authClient.Client): I.UserInfoClient => ({
+  findUserByFederationToken: findUserBySession(client),
+});
+
+const makeIOBackendClient = (
+  baseUrl: string,
```
Comment (datalek) [inline src/userinfo/ioUserInfoClient.ts:64]: What do you think to type baseUrl to `URL` instead of `string`?

File: src/userinfo/ioUserInfoClient.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  fetchApi: typeof fetch = nodeFetch as unknown as typeof fetch
```
Comment (datalek) [inline src/userinfo/ioUserInfoClient.ts:65]: :goberserk:

File: src/userinfo/__tests__/ioUserInfoClient.test.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+describe("makeIOUserInfoClient", () => {
+  describe("findUserByFederationToken", () => {
+    it("should return a user info", async () => {
+      const { service, mockClient } = p.makeService();
+
+      const functionRecorded = mockClient.getUserIdentity.mockReturnValueOnce(
+        Promise.resolve(
+          t.success({
+            status: 200,
```
Comment (datalek) [inline src/userinfo/__tests__/ioUserInfoClient.test.ts:16]: In your opinion should we move also this `struct` into `phonies` or `records` files?

File: .env.default (hunk 1 above)
Comment (kin0992) [reply .env.default:2]: It's an error. I'm going to put back the original value.

File: .env.default (hunk 2 above)
Comment (kin0992) [reply .env.default:8]: It's an error. I'm going to put back the original value.

File: src/oidcprovider/__tests__/router.test.ts (hunk 3 above)
Comment (kin0992) [reply src/oidcprovider/__tests__/router.test.ts]: Done!

File: src/userinfo/__tests__/ioUserInfoClient.test.ts (hunk 7 above)
Comment (kin0992) [reply src/userinfo/__tests__/ioUserInfoClient.test.ts:16]: Yeah, but can we keep here for now?

File: src/userinfo/ioUserInfoClient.ts (hunk 4 above)
Comment (kin0992) [reply src/userinfo/ioUserInfoClient.ts]: Agree! Done!

File: src/userinfo/ioUserInfoClient.ts (hunk 6 above)
Comment (kin0992) [reply src/userinfo/ioUserInfoClient.ts]: 😂

File: src/userinfo/ioUserInfoClient.ts (hunk 5 above)
Comment (kin0992) [reply src/userinfo/ioUserInfoClient.ts]: I agree!

**Comment by datalek** [comment]
👍

**Review by kin0992 — APPROVED** [review]
